hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieFileSliceReader.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.common.table.log;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.generic.IndexedRecord;
+import org.apache.hudi.common.model.HoodieRecord;
+import org.apache.hudi.common.model.HoodieRecordPayload;
+import org.apache.hudi.common.util.SpillableMapUtils;
+import org.apache.hudi.io.storage.HoodieFileReader;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Iterator;
+
+/**
+ * Reads records from base file and merges any updates from log files and provides iterable over all records in the file slice.
+ */
+public class HoodieFileSliceReader implements Iterable<HoodieRecord<? extends HoodieRecordPayload>> {
+  private HoodieMergedLogRecordScanner logRecordScanner;
+
+  public static <R extends IndexedRecord, T extends HoodieRecordPayload> HoodieFileSliceReader getFileSliceReader(
+      HoodieFileReader<R> baseFileReader, HoodieMergedLogRecordScanner scanner, Schema schema, String payloadClass) throws IOException {
```

Comment:
Should we make this AbstractLogRecordScanner to allow for UnMergedScanner as well ? @satishkotha

Reply:
UnmergedScanner doesnt seem to have a way to get record iterator. This requires lot more refactoring if we want to make it work for both.

Reply:
Hmm, okay, in that case, can we keep the method signature of FileSliceReader as iterator rather than scanner so that way the FileSliceReader is generic and then we can figure out how to make those record scanner changes later ?
@satishkotha

Reply:
We also need process method in LogRecordScanner. So I kept this 'factory' method as is. But changed HoodieFileSliceReader constructor to take records. We can implement another factory method for UnmergedScanner. PTAL.

Reply:
@satishkotha We cannot use the scanner.getRecords() API, that will bring all the records into memory and cause the job to OOM. Let's just continue to return the iterator and then lazily fetch records as you iterate.

Reply:
@n3nash getRecords just retruns 'ExternalSpillableMap'. So i dont think it can cause OOM. Anyway, changed it to iterator to keep it similar to reading records from base files.

Reply:
Nonetheless we should file a `code cleanup` JIRA to provide these iterators as core building blocks under a nice abstractions.